CodeCompletion offers non-static method via "using static Class;"

![image](https://cloud.githubusercontent.com/assets/774791/11458740/4d08c410-96c6-11e5-8bde-147084bb5039.png)


Program.cs:
```csharp
using static ClassWithNonStaticMethod;
class Program
{
    static void Main(string[] args)
    {
        var instace = new Program();
        instace.NonStaticMethod();
    }

    private void NonStaticMethod()
    {
        MathMin(0, 1);
    }
}
```

ClassWithNonStaticMethod.cs:
```csharp
using System;

class ClassWithNonStaticMethod
{
    public static int MathMax(int a, int b)
    {
        return Math.Max(a, b);
    }

    public int MathMin(int a, int b)
    {
        return Math.Min(a, b);
    }
}
```

Fun fact: Bug was discovered while using Xamarin Studio :)

A revision of the language spec that I have says the following: "A *using-static-directive* imports the nested types and **static members** contained directly in a type declaration into the immediately enclosing compilation unit or namespace body, enabling the identifier of each member and type to be used without qualification."

Also, compiler reports 
```
error CS0103: The name 'MathMin' does not exist in the current context
```
in this scenario.

It looks to me that Semantic model behavior is inconsistent with the spec and with lookup rules used by  the binder.

--- a/src/Compilers/CSharp/Portable/Binder/Imports.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Imports.cs
@@ -672,34 +672,9 @@ namespace Microsoft.CodeAnalysis.CSharp
                 ImmutableArray<Symbol> candidates = Binder.GetCandidateMembers(typeOrNamespace.NamespaceOrType, name, options, originalBinder: originalBinder);
                 foreach (Symbol symbol in candidates)
                 {
-                    switch (symbol.Kind)
+                    if (!IsValidLookupCandidateInUsings(symbol))
                     {
-                        // lookup via "using namespace" ignores namespaces inside
-                        case SymbolKind.Namespace:
-                            continue;
-
-                        // lookup via "using static" ignores extension methods and non-static methods
-                        case SymbolKind.Method:
-                            if (!symbol.IsStatic || ((MethodSymbol)symbol).IsExtensionMethod)
-                            {
-                                continue;
-                            }
-
-                            break;
-
-                        // types are considered static members for purposes of "using static" feature
-                        // regardless of whether they are declared with "static" modifier or not
-                        case SymbolKind.NamedType:
-                            break;
-
-                        // lookup via "using static" ignores non-static members
-                        default:
-                            if (!symbol.IsStatic)
-                            {
-                                continue;
-                            }
-
-                            break;
+                        continue;
                     }
 
                     // Found a match in our list of normal using directives.  Mark the directive
@@ -714,6 +689,41 @@ namespace Microsoft.CodeAnalysis.CSharp
                     result.MergeEqual(res);
                 }
             }
+        }
+
+        private static bool IsValidLookupCandidateInUsings(Symbol symbol)
+        {
+            switch (symbol.Kind)
+            {
+                // lookup via "using namespace" ignores namespaces inside
+                case SymbolKind.Namespace:
+                    return false;
+
+                // lookup via "using static" ignores extension methods and non-static methods
+                case SymbolKind.Method:
+                    if (!symbol.IsStatic || ((MethodSymbol)symbol).IsExtensionMethod)
+                    {
+                        return false;
+                    }
+
+                    break;
+
+                // types are considered static members for purposes of "using static" feature
+                // regardless of whether they are declared with "static" modifier or not
+                case SymbolKind.NamedType:
+                    break;
+
+                // lookup via "using static" ignores non-static members
+                default:
+                    if (!symbol.IsStatic)
+                    {
+                        return false;
+                    }
+
+                    break;
+            }
+
+            return true;
         }
 
         internal void LookupExtensionMethodsInUsings(
@@ -831,7 +841,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             {
                 foreach (var member in namespaceSymbol.NamespaceOrType.GetMembersUnordered())
                 {
-                    if (originalBinder.CanAddLookupSymbolInfo(member, options, null))
+                    if (IsValidLookupCandidateInUsings(member) && originalBinder.CanAddLookupSymbolInfo(member, options, null))
                     {
                         result.AddSymbol(member, member.Name, member.GetArity());
                     }

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/BindingTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/BindingTests.cs
@@ -3046,5 +3046,113 @@ class C
 
             Assert.Equal("System.Threading.Tasks.Task<AuthenticationResult> AuthenticationManager.AuthenticateAsync(System.String authenticationScheme)", group.Single().ToTestDisplayString());
         }
+
+        [Fact, WorkItem(7101, "https://github.com/dotnet/roslyn/issues/7101")]
+        public void UsingStatic_01()
+        {
+            var source =
+@"
+using System;
+using static ClassWithNonStaticMethod;
+using static Extension1;
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        var instace = new Program();
+        instace.NonStaticMethod();
+    }
+
+    private void NonStaticMethod()
+    {
+        MathMin(0, 1);
+        MathMax(0, 1);
+        MathMax2(0, 1);
+        
+        int x;
+        x = F1;
+        x = F2;
+
+        x.MathMax2(3);
+    }
+}
+
+class ClassWithNonStaticMethod
+{
+    public static int MathMax(int a, int b)
+    {
+        return Math.Max(a, b);
+    }
+
+    public int MathMin(int a, int b)
+    {
+        return Math.Min(a, b);
+    }
+
+    public int F2 = 0;
+}
+
+static class Extension1
+{
+    public static int MathMax2(this int a, int b)
+    {
+        return Math.Max(a, b);
+    }
+
+    public static int F1 = 0;
+}
+
+static class Extension2
+{
+    public static int MathMax3(this int a, int b)
+    {
+        return Math.Max(a, b);
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib45(source);
+
+            comp.VerifyDiagnostics(
+    // (16,9): error CS0103: The name 'MathMin' does not exist in the current context
+    //         MathMin(0, 1);
+    Diagnostic(ErrorCode.ERR_NameNotInContext, "MathMin").WithArguments("MathMin").WithLocation(16, 9),
+    // (18,9): error CS0103: The name 'MathMax2' does not exist in the current context
+    //         MathMax2(0, 1);
+    Diagnostic(ErrorCode.ERR_NameNotInContext, "MathMax2").WithArguments("MathMax2").WithLocation(18, 9),
+    // (22,13): error CS0103: The name 'F2' does not exist in the current context
+    //         x = F2;
+    Diagnostic(ErrorCode.ERR_NameNotInContext, "F2").WithArguments("F2").WithLocation(22, 13)
+                );
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var node1 = tree.GetRoot().DescendantNodes().Where(n => n.IsKind(SyntaxKind.IdentifierName) && ((IdentifierNameSyntax)n).Identifier.ValueText == "MathMin").Single().Parent;
+            Assert.Equal("MathMin(0, 1)", node1.ToString());
+
+            var names = model.LookupNames(node1.SpanStart);
+            Assert.False(names.Contains("MathMin"));
+            Assert.True(names.Contains("MathMax"));
+            Assert.True(names.Contains("F1"));
+            Assert.False(names.Contains("F2"));
+            Assert.False(names.Contains("MathMax2"));
+            Assert.False(names.Contains("MathMax3"));
+
+            Assert.True(model.LookupSymbols(node1.SpanStart, name: "MathMin").IsEmpty);
+            Assert.Equal(1, model.LookupSymbols(node1.SpanStart, name: "MathMax").Length);
+            Assert.Equal(1, model.LookupSymbols(node1.SpanStart, name: "F1").Length);
+            Assert.True(model.LookupSymbols(node1.SpanStart, name: "F2").IsEmpty);
+            Assert.True(model.LookupSymbols(node1.SpanStart, name: "MathMax2").IsEmpty);
+            Assert.True(model.LookupSymbols(node1.SpanStart, name: "MathMax3").IsEmpty);
+
+            var symbols = model.LookupSymbols(node1.SpanStart);
+            Assert.False(symbols.Where(s => s.Name == "MathMin").Any());
+            Assert.True(symbols.Where(s => s.Name == "MathMax").Any());
+            Assert.True(symbols.Where(s => s.Name == "F1").Any());
+            Assert.False(symbols.Where(s => s.Name == "F2").Any());
+            Assert.False(symbols.Where(s => s.Name == "MathMax2").Any());
+            Assert.False(symbols.Where(s => s.Name == "MathMax3").Any());
+        }
     }
 }

